packages/tasit-action/src/TasitAction.test.js

```diff
@@ -84,109 +77,292 @@ describe("Contract", () => {
-    } catch (e) {
-      assert(true);
-    }
-  });
 
-  it("should throw error then listening on invalid event", async () => {
-    const events = ["ValueChanged"];
-    const subscription = simpleStorage.subscribe(events);
+      expect(() => {
+        subscription.on("InvalidEvent", () => {});
+      }).to.throw();
+    });
 
-    try {
-      subscription.on("InvalidEvent", () => {});
-      assert(false, "listening with invalid event");
-    } catch (e) {
-      assert(true);
-    }
-  });
+    it("should listen to an event", async () => {
+      const events = ["ValueChanged"];
+      const subscription = simpleStorage.subscribe(events);
+      const fakeFn = sinon.fake();
 
-  it.skip("should listen to an event", async () => {
-    const events = ["ValueChanged"];
-    const subscription = simpleStorage.subscribe(events);
-    const handlerFunction = async message => {};
+      const handlerFunction = message => {
+        const { data } = message;
+        const { args } = data;
+        fakeFn();
+        subscription.removeListener("ValueChanged", handlerFunction);
+      };
+
+      subscription.on("ValueChanged", handlerFunction);
+
+      simpleStorage.setValue("hello world");
+
+      await mineBlocks(simpleStorage.getProvider(), 10);
+
+      expect(fakeFn.called).to.be.true;
+    });
 
-    subscription.on("ValueChanged", handlerFunction);
+    it("should remove listener after timeout", async () => {
+      const events = ["ValueChanged"];
+      const subscription = simpleStorage.subscribe(events);
+      const eventFn = sinon.fake();
+      const errorFn = sinon.fake();
+
+      const foreverCallback = async message => {
+        eventFn();
+      };
+
+      await subscription.on("ValueChanged", foreverCallback);
+
+      await subscription.on("error", err => {
+        errorFn();
+      });
+
+      simpleStorage.setValue("hello world");
+
+      await mineBlocks(simpleStorage.getProvider(), 10);
+
+      expect(eventFn.called).to.be.true;
+
+      // TODO: Use fake timer
+      await new Promise(resolve => setTimeout(resolve, 3000));
+
+      expect(
+        subscription.listenerCount("ValueChanged"),
+        "listener should be removed after timeout"
+      ).to.be.equals(0);
+
+      expect(errorFn.called).to.be.true;
+    });
+
+    it("should manage many listeners", async () => {
+      const events = ["ValueChanged"];
+      const subscription = simpleStorage.subscribe(events);
+
+      const listener1 = message => {};
+      const listener2 = message => {};
+      const listener3 = message => {};
+
+      expect(subscription.listenerCount("ValueChanged")).to.be.equal(0);
+
+      subscription.on("ValueChanged", listener1);
+      subscription.on("ValueChanged", listener2);
+
+      expect(subscription.listenerCount("ValueChanged")).to.be.equal(2);
+
+      subscription.off("ValueChanged", listener2);
+
+      expect(subscription.listenerCount("ValueChanged")).to.be.equal(1);
```

Comment:
Also, we could have a function listenerCount() with no params that returns listeners across all events, and have a test for that.

Reply:
Method `subscription.eventNames()` was introduced on PR https://github.com/tasitlabs/TasitSDK/pull/102 and it's used on several test cases. Should we create a specific `subscription.eventNames()` test case?

Reply:
Hmm, couldn't hurt, but that's a pretty low priority. Maybe create a `good first issue` for that?